### PR TITLE
Moosedocs float referencing and algorithm extension

### LINKED ISSUE #20201: MooseDocs reference floats from other pages
## Reason
It would be nice to reference tables and figures from other pages, like the `eqref` command.

## Design

Similar to `eqref`, something like `[!ref](page_name.md#float_id)`. Autolink should still work within the same page.

## Impact

Better documentation linking.

### PR DESCRIPTION
**Pull request by zachmprince**
Closes #20201 #20202

### CODE REVIEW

**Comment by loganharbour** [comment]
Is this ready for review? Looks like all of the failures are just golds you need to fix up?

**Comment by zachmprince** [comment]
@loganharbour @cticenhour This should be good to review now. Had to learn how the gold files worked for MooseDocs.

File: python/MooseDocs/extensions/floats.py [hunk 1]
```diff
@@ -107,6 +116,30 @@ def postTokenize(self, page, ast):
                     tokens.String(shortcut, content='{} {}'.format(prefix.title(), node['number']))
 
         page['counts'] = counts
+        page['floats'] = floats
+
+class FloatReferenceCommand(command.CommandComponent):
+    COMMAND = 'ref'
+    SUBCOMMAND = None
+    LABEL_RE = re.compile(r'((?P<filename>.*?\.md)#)?(?P<label>.+)', flags=re.UNICODE)
```
Comment (cticenhour) [inline python/MooseDocs/extensions/floats.py:124]: Is this new syntax being explicitly tested? I see the impact on the changes to floats as a result of this (e.g., changing how `id` is represented in the syntax and the need for re-golding), but I'm not seeing testing of this new syntax. Apologies if I am missing it in the GitHub diff UI.

**Review by cticenhour — CHANGES_REQUESTED** [review]
Really cool new features! Just a few comments.

File: python/MooseDocs/extensions/floats.py [hunk 2]
```diff
@@ -160,10 +193,68 @@ def createHTML(self, parent, token, page):
...
+        if token['filename']:
+            float_page = self.translator.findPage(token['filename'], throw_on_zero=False)
+            if float_page is None:
+                a['class'] = 'moose-error'
+                html.String(a, content='{}#{}'.format(token['filename'], key))
+                msg = "Could not find  page {}".format(token['filename'])
+                raise exceptions.MooseDocsException(msg)
+                # LOG.error(common.report_error(msg, page.source,
+                #                               token.info.line if token.info else None,
+                #                               token.info[0] if token.info else token.text()))
```
[the remark this answers is not in the document]
Comment (zachmprince) [reply python/MooseDocs/extensions/floats.py:218]: Good catch. Thanks.

File: python/MooseDocs/extensions/floats.py (hunk 1 above)
Comment (zachmprince) [reply python/MooseDocs/extensions/floats.py:124]: Shoot, yeah, I need to test and document this.

File: modules/doc/content/python/MooseDocs/extensions/algorithm.md [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+!devel-end!
+
+!devel settings caption=Configuration options for the algorithm command.
+                module=MooseDocs.extensions.algorithm
+                object=AlgorithmCommand
+
+
+## Line Commands
+
+The algorithm command by itself is not very informative, user's should use the following line commands following the algorithm float definition. Below is an example using some of the commands described in the following sections. $\quad$
```
[the remark this answers is not in the document]
Comment (cticenhour) [reply modules/doc/content/python/MooseDocs/extensions/algorithm.md:26]: FYI - made an edit to the above suggestion....realized it should be two sentences.

**Review by cticenhour — APPROVED** [review]
This looks great. Thanks for adding the testing and doco for the new reference syntax!

**Pull request merged**